# Part 2

## Calculating Fst between the two homozygote groups

INPUT: Individual IDs for the two homozygote groups, bcf file of the
chromosome, determined in the genotyping step above.

(Need `vcf_col_ids` which is a list of column names in the vcf file
corresponding to the individuals. Can be generated by
`bcftools query -l NW_022145594.1_filtered.vcf > vcf_col_ids`)

### Step 1: Column IDs in the vcf file belonging to p and q
```bash
#!/bin/bash

file_path="vcf_col_ids"

if [ $2 == 3 ]; then
file_homop="${1}_homop_nums.txt"
file_homoq="${1}_homoq_nums.txt"

homop=()
while IFS= read -r line; do
  homop+=("$line")
done < $file_homop

homoq=()
while IFS= read -r line; do
      homoq+=("$line")
    done < $file_homoq

for row_number in "${homop[@]}"; do
  sed -n "${row_number}p" "$file_path" >> "${1}_vcf_list_homop"
done

for row_number in "${homoq[@]}"; do
    sed -n "${row_number}p" "$file_path" >> "${1}_vcf_list_homoq"
done

else

for c in $(seq 0 $2)
do
  echo "number $c"
  myfile="${1}_clust${c}.txt"
  geno=()
  while IFS= read -r line; do
    geno+=("$line")
  done < $myfile
  for row_number in "${geno[@]}"; do
    sed -n "${row_number}p" "$file_path" >> "${1}_vcf_list_clust${c}"
  done

done
fi
```

`chr=NW_022145594.1`
`fname=NW_022145594.1_39429440_42445994`

run: `bash extract_inds_from_vcf.sh ${fname} 4`, 4 is number of clusters minus 1 (except if it's 3, then write 3), output is NW_022145594.1_12670717_16440127_vcf_list_homop and
NW_022145594.1_12670717_16440127_vcf_list_homoq for 3 clusters or clust0...4 for 5 clusters which are lists of vcf column names to keep for each group

Resulting files are in the intermediary_files directory.

### Step 2: Run Fst calculation

`spack load vcftools@0.1.14`

`vcftools --gzvcf ~/EG2023/structural_variation/backup/filtered_vcf/NW_022145594.1_filtered.vcf --weir-fst-pop NW_022145594.1_12670717_16440127_vcf_list_homoq --weir-fst-pop NW_022145594.1_12670717_16440127_vcf_list_homop --out NW_022145594.1_12670717_16440127_fst`

`vcftools --gzvcf ~/EG2023/structural_variation/backup/filtered_vcf/${chr}_filtered.vcf --weir-fst-pop ${fname}_vcf_list_homoq --weir-fst-pop ${fname}_vcf_list_homop --out ${fname}_fst`

### Step 3: Plot Fst results

```python
import pandas as pd
import matplotlib.pyplot as plt
import argparse

parser = argparse.ArgumentParser(description="Description of your script")
parser.add_argument("chr", type=str, help="Chromosome name")
parser.add_argument("start", type=int, help="start region")
parser.add_argument("stop", type=int, default=0.1, help="stop region")
args = parser.parse_args()

filename="~/WGS/inversion_results/" + args.chr + "_" + str(args.start) + "_" + str(args.stop) + "_fst.weir.fst"
df=pd.read_csv(filename, sep="\t")

fig, ax1 = plt.subplots(figsize=(16, 8))
plt.plot(df["POS"], df["WEIR_AND_COCKERHAM_FST"], ".")
plt.ylabel("Fst, homo1 vs homo2")
plt.xlabel("Genomic position")

plt.savefig(args.chr + "_" + str(args.start) + "_" + str(args.stop) + "fst_plot.jpg")
```

`python fst_plotting.py NW_022145594.1 12670717 16440127`

OUTPUT: Fst value for each position in the chromosome, and Fst plot

Resulting Fst files are on Zenodo in the fst_files directory. TODO
Resulting plots (for inversions 1 and 6) are in the intermediary_files directory.

## Linkage disequilibrium

### Step 1: Chromosome vcf file to gds

conda activate grn
Run:
`Rscript vcf2gds.R ${filtered}/NW_022145594.1_filtered.vcf NW_022145594.1`,
same script from before, makes gds file

### Step 2: Calculate LD with a certain window size

To check how many windows a specific window size will produce run [LD_get_num_windows.R](https://github.com/Cpetak/Urchin_inversions/blob/main/LD_get_num_windows.R).

`Rscript LD_get_num_windows.R NW_022145594.1.gds 50000`

***NOTE:*** Around 1,000 windows is good. Too big window size could break the code if there are less than 1 windows, but too small window size will take literally ages to compute!

In our case, with 50_000 bp we get 1061 windows.

`mkdir makegrid_NW_022145594.1_50000` this is where the small temporary files will be saved. Run [makegrid_launcher.sh](https://github.com/Cpetak/Urchin_inversions/blob/main/makegrid_launcher.sh) which creates a job array that runs [makegrid.R](https://github.com/Cpetak/Urchin_inversions/blob/main/makegrid.R).

run: `sbatch makegrid_launcher.sh ~/WGS/inversion_results/NW_022145594.1.gds 50000`, file name, window size, submits 200 jobs in the short partition. Output is 200 files, each containing a subset of the windows.

### Step 3: Put together the set of files

Run [gather_results.R](https://github.com/Cpetak/Urchin_inversions/blob/main/gather_results.R), `Rscript gather_results.R NW_022145594.1 50000` , output is combined_chrom_windowsize.Rdata

Free up space: `rm -r makegrid_NW_022145594.1_50000`

Combined LD files can be found in Zenodo ld_data directory. TODO

### Step 4: Make triangle plot

Run [plot_LD.R](https://github.com/Cpetak/Urchin_inversions/blob/main/plot_LD.R), `Rscript plot_LD.R NW_022145594.1 50000`, output is a pdf.

LD plots for each inversion can be found in the intermediary_files directory.

### Triangle plot for specific region

For zooming in a specific region, repeat above but with a subsetted vcf

***NOTE:*** Above code can be used the same but file names only include chromosome and LD window size so make sure to move output figures to avoid overwriting them!

For region 12702886 - 16793794 plus padding: 8702886 - 20793794. 

`Rscript ../local_pca_pipe/LD_get_num_windows.R NW_022145594.1_8702886_20793794.gds 10000` -> 1208

### U-plot:

For the U-plot, I calculated the LD for each pair of SNPs, in the 12702886 - 16793794 range, only keeping r2>0.6 to reduce the file size.

```bash
sed 's/_//g' NW_022145594.1_12702886_16793794.vcf > ld_input_NW_022145594.1_12702886_16793794.vcf
./plink --vcf ld_input_NW_022145594.1_12702886_16793794.vcf --recode --allow-extra-chr --out NW_022145594.1_12702886_16793794_plink
./plink --file NW_022145594.1_12702886_16793794_plink --make-bed --allow-extra-chr --out NW_022145594.1_12702886_16793794_afterQC
./plink --bfile NW_022145594.1_12702886_16793794_afterQC --r2 --allow-extra-chr --ld-window-r2 0.6 --ld-window 100000000 --ld-window-kb 100000000 --out NW_022145594.1_12702886_16793794_resultLD
```

Note:
```bash
--ld-window 10 #compare 2 SNPS only if they are within 10 SNPS apart
--ld-window-kb 1000 #compare 2 SNPS only if they are within 1.000.000 bp apart (both needs to be true)
--ld-window-r2 0 #output any LD value, regardless of what it is. default would be 0.2, so only values that are above 0.2 are written to file
```

## Nucleotide diversity

For each chromosome, need to be done only once:

```bash
filtered=/users/c/p/cpetak/EG2023/structural_variation/filtered_bcf_files/NW_022145594.1
# spack load vcftools@0.1.14
vcftools --vcf $filtered/NW_022145594.1_filtered.vcf --site-pi --out NW_022145594.1_nuc_div
```

Output: Pi nucleotide diversity for each site.

`vcftools --vcf ${filtered}/NW_022145594.1_filtered.vcf --window-pi 10000 --out NW_022145594.1_nuc_div` 

Output: Pi average over windows.

`vcftools --vcf ${filtered}/NW_022145594.1_filtered.vcf --TajimaD 10000 --out NW_022145594.1_nuc_div`

Output: Tajima's D over windows.

### Zoom in and plot homozygotes only and q and p separately

For example, region: 12.702.886-16.793.794, expanded by 100,000 both ends -> 12.602.886 - 16.893.794

1. Subset vcf to extended region of interest, like in Part 1. 

2. Get only homozygotes to one orientation:

```bash
#spack load bcftools@1.10.2
bcftools view -S NW_022145594.1_12702886_16793794_vcf_list_homop -o NW_022145594.1_12602886_16893794_homop.vcf NW_022145594.1_12602886_16893794.vcf
bcftools view -S NW_022145594.1_12702886_16793794_vcf_list_homoq -o NW_022145594.1_12602886_16893794_homoq.vcf NW_022145594.1_12602886_16893794.vcf
```
```bash
chr="NW_022145594.1" start=12702886 stop=16793794
myfolder=~/EG2023/structural_variation/filtered_bcf_files/${chr}
bcftools view -S ${chr}_${start}_${stop}_vcf_list_homop -o ${chr}_${start}_${stop}_homop.vcf ${myfolder}/${chr}_filtered.vcf
bcftools view -S ${chr}_${start}_${stop}_vcf_list_homoq -o ${chr}_${start}_${stop}_homoq.vcf ${myfolder}/${chr}_filtered.vcf
```

3. Calculate nucleotide diversity like above, smaller window size.

```bash
vcftools --vcf ${chr}_${start}_${stop}_homop.vcf --window-pi 100 --out ${chr}_${start}_${stop}_nuc_div_homop
vcftools --vcf ${chr}_${start}_${stop}_homoq.vcf --window-pi 100 --out ${chr}_${start}_${stop}_nuc_div_homoq

vcftools --vcf ${chr}_${start}_${stop}_homop.vcf --TajimaD 100 --out ${chr}_${start}_${stop}_nuc_div_homop
vcftools --vcf ${chr}_${start}_${stop}_homoq.vcf --TajimaD 100 --out ${chr}_${start}_${stop}_nuc_div_homoq
```

Adding 100,000 was not enough, doing same as above but with whole chromosome instead (then zooming in at plotting)

```bash
bcftools view -S NW_022145594.1_12702886_16793794_vcf_list_homop -o NW_022145594.1_homop.vcf ${filtered}/NW_022145594.1_filtered.vcf
bcftools view -S NW_022145594.1_12702886_16793794_vcf_list_homoq -o NW_022145594.1_homoq.vcf ${filtered}/NW_022145594.1_filtered.vcf
vcftools --vcf NW_022145594.1_homop.vcf --window-pi 10000 --out NW_022145594.1_nuc_div_homop
vcftools --vcf NW_022145594.1_homoq.vcf --window-pi 10000 --out NW_022145594.1_nuc_div_homoq
vcftools --vcf NW_022145594.1_homoq.vcf --TajimaD 10000 --out NW_022145594.1_nuc_div_homoq
vcftools --vcf NW_022145594.1_homop.vcf --TajimaD 10000 --out NW_022145594.1_nuc_div_homop
```

Nucleotide diversity data is available on Zenodo nuc_div_data directory TODO.
For plotting nucleotide diversities: [nuc_div.ipynb](https://github.com/Cpetak/Urchin_inversions/blob/main/nuc_div.ipynb)
**Figure 2, supplementary figure of nucleotide diversities and Figure 5 subplot with PI diversity and SNP count are all plotted here!**



